draft-irtf-cfrg-opaque.md

```diff
@@ -398,145 +398,331 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
 
-Both client and server MUST validate the other party's public key before use.
-See {{validation}} for more details.
+## Envelope Structure {#envelope-structure}
 
-Upon completion, S stores C's credentials for later use. See {{credential-file}}
-for a recommended storage format.
+A client `Envelope` is constructed independently from the `EnvelopeMode`, but offers an `InnerEnvelope` entry that modes can use as they see fit.
+`Envelope` is constructed as follows:
+~~~
+struct {
+ EnvelopeMode mode;
+ opaque nonce[Nn];
+ opaque auth_tag[Nh];
+ InnerEnvelope inner_env;
+} Envelope;
+~~~
 
-## Credential Storage {#credential-storage}
+mode : The `EnvelopeMode` used for the `Envelope`.
 
-OPAQUE makes use of a structure `Envelope` to store client credentials.
-The `Envelope` structure embeds the following types of credentials:
+nonce : A unique nonce of length `Nn` used to protect this Envelope.
 
-- client_private_key: The encoded client private key for the AKE protocol.
-- server_public_key: The encoded server public key for the AKE protocol.
-- client_identity: The client identity. This is an application-specific value, e.g., an e-mail
-  address or normal account name.
-- server_identity: The server identity. This is typically a domain name, e.g., example.com.
-  See {{identities}} for information about this identity.
+auth_tag : Authentication tag protecting the contents of the envelope, covering `InnerEnvelope` and `CleartextCredentials`.
+
+inner_env : A mode dependent `InnerEnvelope` specifying at least the `EnvelopeMode`.
+
+The `CreateEnvelope` function allows a client to create its `Envelope` at registration.
+~~~
+CreateEnvelope(y, pks, creds)
+
+Parameters:
+- params, the MHF parameters established out of band
+- mode, the EnvelopeMode mode
+- Nx, the output size of the Extract function
+- Nn, the nonce length
+
+Input:
+- y, finalized OPRF output
+- creds, a Credentials structure
+- pks, the server AKE public key
+
+Output:
+- envelope, the client's `Envelope` structure
+- client_public_key, the client's AKE public key
+- export_key, an additional key
+
+Steps:
+1. envelope_nonce = random(Nn)
+2. prk = Extract(envelope_nonce, Harden(y, params))
+3. auth_key = Expand(prk, "AuthKey", Nx)
+4. export_key = Expand(prk, "ExportKey", Nx)
+5. inner_env, client_public_key = BuildInnerEnvelope(prk, creds)
+6. cleartext_creds = CreateCleartextCredentials(response.server_public_key, client_public_key, creds.server_identity, creds.client_identity)
+7. auth_tag = MAC(auth_key, concat(mode, envelope_nonce, inner_env, cleartext_creds))
+8. Create Envelope envelope with (mode, envelope_nonce, auth_tag, inner_env)
+9. Output (envelope, client_public_key, export_key)
+~~~
+
+The `RecoverEnvelope` function allows a client to recover its `Envelope` at authentication. Note that the `creds` structure MUST at least hold the server_public_key and client_public_key values.
```

Comment:
The server public key is passed in as a parameter (as part of the CredentialResponse), so why is it needed to be part of `creds`?

Reply:
Indeed, I think this is a leftover of a drafting phase. In my implementation, I only use this field to pass the server public key to the server when responding on login. Thanks for spotting this !

Reply:
And that I'm thinking of it, client_public_key isn't necessary either.

Reply:
Fixed.